aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,218 @@
 # -*- coding: utf-8 -*-
-"""`CalcJob` implementation for the pp.x code of Quantum ESPRESSO."""
+""" CalcJob implementation for pp.x """
 from __future__ import absolute_import
 
-from aiida.orm import RemoteData, FolderData
-from aiida_quantumespresso.calculations.namelists import NamelistsCalculation
+import os
+import six
+from six.moves import range
 
+from aiida import orm
+from aiida.common import datastructures, exceptions
 
-class PpCalculation(NamelistsCalculation):
+from aiida_quantumespresso.calculations import _lowercase_dict, _uppercase_dict
+from aiida_quantumespresso.utils.convert import convert_input_to_namelist_entry
+from .base import CalcJob
+
+
+class PpCalculation(CalcJob):
     """`CalcJob` implementation for the pp.x code of Quantum ESPRESSO."""
 
+    # Default name of the subfolder inside 'parent_folder' from which you want to copy the files, in case the
+    # parent_folder is of type FolderData
+    _INPUT_SUBFOLDER = './out/'
+
+    # In the PW Calculation plugin, these folder names and prefixes are fixed, so import them to reduce maintenance
+    # pylint: disable=protected-access
+    from aiida_quantumespresso.calculations import BasePwCpInputGenerator
+    _OUTPUT_SUBFOLDER = BasePwCpInputGenerator._OUTPUT_SUBFOLDER
+    _PREFIX = BasePwCpInputGenerator._PREFIX
+    _PSEUDO_SUBFOLDER = BasePwCpInputGenerator._PSEUDO_SUBFOLDER
+    _DEFAULT_INPUT_FILE = BasePwCpInputGenerator._DEFAULT_INPUT_FILE
+    _DEFAULT_OUTPUT_FILE = BasePwCpInputGenerator._DEFAULT_OUTPUT_FILE
+    # pylint: enable=protected-access
+
+    # Grid data output file from first stage of pp calculation
     _FILPLOT = 'aiida.filplot'
+    # Grid data output in desired format
+    _FILEOUT = 'aiida.fileout'
 
     _default_namelists = ['INPUTPP', 'PLOT']
-    _internal_retrieve_list = [_FILPLOT]
-    _blocked_keywords = [
-        ('INPUTPP', 'outdir', NamelistsCalculation._OUTPUT_SUBFOLDER),
-        ('INPUTPP', 'prefix', NamelistsCalculation._PREFIX),
-        ('INPUTPP', 'filplot', _FILPLOT),
-    ]
+    _internal_retrieve_list = [_FILPLOT, _FILEOUT]
+
+    # Keywords that cannot be set by the user but will be set by the plugin
+    _blocked_keywords = [('INPUTPP', 'outdir', _OUTPUT_SUBFOLDER), ('INPUTPP', 'prefix', _PREFIX),
+                         ('INPUTPP', 'filplot', _FILPLOT), ('PLOT', 'fileout', _FILEOUT),
+                         ('PLOT', 'output_format', None)]
+
+    _retrieve_singlefile_list = []
 
     @classmethod
     def define(cls, spec):
         # yapf: disable
         super(PpCalculation, cls).define(spec)
-        spec.input('parent_folder', valid_type=(RemoteData, FolderData),
+        spec.input('parent_folder', valid_type=(orm.RemoteData, orm.FolderData), required=True,
             help='Output folder of a completed `PwCalculation`')
+        spec.input('parameters', valid_type=orm.Dict, required=False,
+            help='Use a node that specifies the input parameters for the namelists')
```

Comment:
Why is this not required even if the validation in `prepare_for_submission` seems to make it out as though it is required. I think it probably should be required, no?

Reply:
Yep, good spot